### ISSUE TITLE
Atypical keyboard shortcuts are not supported

### ISSUE THREAD

**Issue by truhlikfredy** [Microsoft/vscode#28576]
- VSCode Version: 1.13.0
- OS Version: amd64 Debian 8 

Steps to Reproduce:

1. Have ThinkPad laptop with the next/previous close to the cursor keys
2. Have custom xmodmapings to map them to F14 and F15 keys

> keycode 166=F14
> keycode 167=F15

3. Try to assign these keys to some functions (I get "unknown" instead of naming the shortcut and I can't assign it to the action)

**Comment by truhlikfredy** [comment]
I think the "unknown" message is the KeyCode.Unknown defined in vs/base/common/keyCodes. But I'm not sure what logic is setting it up, could it be some mismatch with scan codes?

**Comment by alexandrudima** [comment]
@truhlikfredy Could you please try our Keybinding troubleshooting guide -- https://github.com/Microsoft/vscode/wiki/Keybindings#troubleshoot-linux-keybindings

**Comment by truhlikfredy** [comment]
Solved it with workardound in the settings.

> "keyboard.dispatch": "keyCode"

But if it's reading original scanCodes, then it should work as well, originally the key are multimedia back / forward (code 166 and 167). Often browsers react to it going to previous website, and actually that was the reason why i remapped it, because I didn't like the original behavior applications had with these keys.

**Comment by alexandrudima** [comment]
Let's continue tracking in #23991 in this case.

**Comment by truhlikfredy** [comment]
#23991 case is for remaps, but in my case it's 2 things, if it would accept the original keycode even ignoring the remaping would be grand as well. If I would remove the remappings, still the visual studio doesn't recognize back/forward keys.

**Comment by alexandrudima** [comment]
I think we recognize F keys up to F19.

**Comment by truhlikfredy** [comment]
Yes, but that is the mapped value, original is this two keys

[https://c1.staticflickr.com/8/7277/6863731296_af6a4baf78_b.jpg](https://c1.staticflickr.com/8/7277/6863731296_af6a4baf78_b.jpg)

**Comment by alexandrudima** [comment]
I can look into adding support in VSCode for them. Can you please press them at https://jsfiddle.net/v1gz5c5m/1/ in Chromium and let me know what values each one has.

**Comment by truhlikfredy** [comment]
This are the original keys

keydown --- code: BrowserForward, keyCode: 167, key: BrowserForward
keydown --- code: BrowserBack, keyCode: 166, key: BrowserBack

**Comment by truhlikfredy** [comment]
Can you remove the duplicate tag? Even if this problem is bit overlapping with other issues, but still this one is unique.

**Comment by antcodd** [comment]
How exactly did you manage to work around this issue? I don't see any way in keybindings.json to specify a keyCode or code and the keybindings editor shows unknown.

The current keybinding design in vscode seems limiting. Why doesn't it just support the names from "code" when keyboard.dispatch=code mode?

Changing keyboard.dispatch to keyCode didn't seem to make a difference for me.

Somewhat related issue: #8641, the same issue for the mouse button equivalent (which I don't believe send keycodes on Linux).

**Comment by truhlikfredy** [comment]
Yes that what I said it works with me because I used xmodmap to map them to F14 and F15 and then added this to vscode settings: 

> "keyboard.dispatch": "keyCode"

Just because I have workaround it doesn't mean it's not an issue so that's why I asked to remove the "duplicate tag" etc...

I looked at the source code and the keyboard management seems too complicated without much gain. Plus now it doesn't respect the remaps which is probably the first application I have seen and has issues with this.

To debug what you get from your linux I like to use 

> xev

It shows exactly what events you are emitting (keyboard, mouse etc...)

**Comment by alexandrudima** [comment]
Keyboard management is complicated because it needs to work correctly in:
* VSCode on Windows (based on Chromium + `native-keymap` to present correct labels to the UI) -- uses `keyCode` to act the same as all other Windows applications
* VSCode on Linux (based on Chromium + `native-keymap` to bind our default US based keybindings to something different depending on the keyboard layout and to present correct labels in the UI) -- uses `code` by default
* VSCode on OSX (based on Chromium + `native-keymap` to bind our default US based keybindings to something different depending on the keyboard layout and to present correct labels in the UI) -- uses `code` by default
* Monaco Editor on Chrome on Windows/OSX/Linux -- uses `keyCode`
* Monaco Editor on Firefox on Windows/OSX/Linux -- uses `keyCode`
* Monaco Editor on IE11 and Edge on Windows -- uses `keyCode`
* Monaco Editor on Safari on OSX -- uses `keyCode`

**Comment by truhlikfredy** [comment]
Wow, that is impresive, didn't know the Monaco Editor is based on this as well, then not really surprised of the over-complications. It needs to be universal and consider many different combinations.

**Comment by jomel** [comment]
Just to say I also have this exact issue.
My keyboard has standalone keys for browser Back and browser forward, that I wanted to use for code folding, but they come up as "unknown" in vscode.
I used the workaround described by @truhlikfredy to assign my keys for F13 & F14.
For anybody else facing the same (on Ubuntu), here is the workaround:

- In vscode open key bindings (ctrl P & "leyboard shortcuts"), 
- click the "record keys" button the right of the input box
- press your key. it will sau "unknown" in the box, but if you hover over that word ift will say something like "code: BrowserBack", keyCode:166, key: "
- in terminal run: `xmodmap -pke > ~/xmodmap_original` : this backup up your current keymap.
- in terminal run: `echo "keycode 166 = F13 F13 F13" => ~/.Xmodmap` : this maps your new key binding
- (for good measure) run `xmodmap ~/.Xmodmap`
- In vscode open settings (ctrl P & "settings", choose "open settings JSON") & add `"keyboard.dispatch": "keyCode"`
- youshould now be able to go back to the vscode key bindings editor & use your new key

**Comment by AntonKrug** [comment]
It stopped working on me recently as it looks like vscode update reset some settings. I forgot I was making this issue and by accident googled my own solution, in this case the "keyboard.dispatch" got reset to default value so I had to set it back to the "keyCode" value.

I have modified other editors, even rebuild chromium from scratch to support the F13/F14 as previous and next tabs, physically these buttons are located close to the cursor arrows and it's very pleasant to have editors/browsers setup so the whole UI is consistent and productive.
